### DISCUSSION TITLE
[QUESTION] Would it make sense to have in-memory SLMs ?

### DISCUSSION THREAD

**Discussion by agoncal** [langchain4j/langchain4j#1411] [Ideas]
Today there are more and more SLMs (_Small Language Models_) that we can use in small devices (Phi-3 mini, Gemini Nano, etc.). Some of them use ONNX files that are smaller than 1 or 2 Gb. 

As LangChain4j already supports in-memory embedding models with ONNX runtimes (see https://github.com/langchain4j/langchain4j-embeddings), what about having a separate repo (ex. `langchain4j-slms`) with a few SLMs so we could easily use them in-memory like we already do with embedding models.

Don't know how feasible this is, but it could be handy for developing Java apps that need SLMs.

WDYT ?

**Comment by LizeRaes** [comment]
Sounds very useful to me, and feasible as well. For fast PoCs this would be amazing, and for real applications that have no huge LLM requirements, it can also be used for the final product.

If we implement this, we might add 1 of 2 defaults, but we could also make it flexible to download any model? This would free us from the task of implementing a version for every new interesting model that comes out.

The trend for smaller and better models will continue, so this is definitely going to be a common use case in the future, and it's a nice way to run a local model (data privacy) on any device without enforcing eg. ollama installation.

**Comment by glaforge** [comment]
I used Gemma 2b within Ollama in Testcontainers: 
https://glaforge.dev/posts/2024/04/04/calling-gemma-with-ollama-and-testcontainers/

It worked great 👍

**Comment by agoncal** [comment]
@glaforge yes, I also use Testcontainers with Ollama. But I was thinking of other use cases. You deploy a Java app in a constrained device (with no Docker) and just run the SLM in memory. So, not just for testing purpose but also for runtime purpose

**Comment by agoncal** [comment]
@langchain4j side question. I am not experienced in ONNX vs GGUF format, but why not supporting GGUF format in https://github.com/langchain4j/langchain4j-embeddings then ?

**Comment by glaforge** [comment]
@agoncal there's also Jlama: https://github.com/tjake/Jlama
